elastic/synthetics-recorder
# feat: Add soft delete to steps functionality

## Summary

Resolves #242.

## Implementation details

Adds an `isSoftDeleted` flag to the `ActionContext` type that we use to represent custom IR values. When a user clicks the `Delete` button while recording, the recorder will apply this flag to the targeted action. ~After recording ends, the recorder will filter that action out of the list.~ Additionally, we will set the `modified` flag on any deleted action. This will cause the existing [`generator` code](https://github.com/elastic/synthetics-recorder/pull/247/files#diff-7cc5d95f5c01783f59eac8f7333a807ce5a224ce3d94d0899aba6d531e120b80R100) to preserve the deleted state when Playwright sends new actions. We need to keep these `modified` actions because even after the Recorder's internal state reflects that recording has stopped, it can still handle latent Playwright events that would overwrite the state; because of this, we can't reliably modify the state to remove the items where `isSoftDeleted == true`.

The export flyout also checks for this flag when generating code. ~This is worth some additional examination, it might be better to modify the code that actually calls the main thread to perform this check on its own, so that we can remove this added abstraction from the UI component layer where possible.~

## How to validate this change

1. Start recording a journey
2. "Delete" an action
3. The action element should disappear
4. Open the export flyout, you should see no code associated with the action you deleted
5. Stop recording and log the steps in `App.tsx`, you should not see any of the actions you soft-deleted while recording
